digest/src/dyn_digest.rs

```diff
@@ -18,33 +18,28 @@ pub trait DynDigest {
     /// Retrieve result and consume boxed hasher instance
     fn finalize(self: Box<Self>) -> Box<[u8]>;
 
-    /// Reset hasher instance to its initial state.
-    fn reset(&mut self);
-
     /// Get output size of the hasher
     fn output_size(&self) -> usize;
 
     /// Clone hasher state into a boxed trait object
     fn box_clone(&self) -> Box<dyn DynDigest>;
 }
 
-impl<D: Update + FixedOutput + Reset + Clone + 'static> DynDigest for D {
+impl<D: Update + FixedOutput + Clone + 'static> DynDigest for D {
     fn update(&mut self, data: &[u8]) {
         Update::update(self, data);
     }
 
+    #[inline]
     fn finalize_reset(&mut self) -> Box<[u8]> {
-        let res = self.clone().finalize_fixed().to_vec().into_boxed_slice();
-        Reset::reset(self);
-        res
-    }
-
-    fn finalize(self: Box<Self>) -> Box<[u8]> {
-        self.finalize_fixed().to_vec().into_boxed_slice()
+        self.clone()
+            .finalize_fixed_reset()
+            .to_vec()
+            .into_boxed_slice()
     }
 
-    fn reset(&mut self) {
-        Reset::reset(self);
+    fn finalize(mut self: Box<Self>) -> Box<[u8]> {
```

Comment:
@newpavlov out of curiosity, why does this return a boxed slice rather than a `Vec<u8>`?

Reply:
`Vec<u8>` takes 3 words, while `Box<[u8]>` only two.

Reply:
Interesting!